langchain/vectorstores/matching_engine.py

```diff
@@ -0,0 +1,465 @@
+        using this module.
+
+        See usage in
+        docs/modules/indexes/vectorstores/examples/matchingengine.ipynb.
+
+        Note that this implementation is mostly meant for reading if you are
+        planning to do a real time implementation. While reading is a real time
+        operation, updating the index takes close to one hour.
+
+        Attributes:
+            project_id: The GCS project id.
+            region: The default location making the API calls. It must have
+            the same location as the GCS bucket.
+            gcs_bucket_uri: The location where the vectors will be stored in
+            order for the index to be created.
+            index: The created index class. See
+            ~:func:`MatchingEngine.from_components`.
+            endpoint: The created endpoint class. See
+            ~:func:`MatchingEngine.from_components`.
+            credentials (Optional): Created GCP credentials.
+            embedding: A :class:`Embeddings` that will be used for
+            embedding the text sent. If none is sent, then the
+            multilingual Tensorflow Universal Sentence Encoder will be used.
+        """
+        super().__init__()
+        self._validate_google_libraries_installation()
+        self.gcs_bucket_uri = self._validate_gcs_bucket(gcs_bucket_uri)
+        self._init_aiplatform(project_id, region, self.gcs_bucket_uri)
+
+        self.project_id = project_id
+        self.region = region
+        self.gcs_client = None
+        self.embedding = embedding
+        self.credentials = credentials
+        self.index = index
+        self.endpoint = endpoint
+
+    def _validate_google_libraries_installation(self) -> None:
+        """Validates that Google libraries that are needed are installed."""
+        try:
+            from google.cloud import aiplatform
+            from google.cloud import storage
+            from google.oauth2 import service_account
+        except ImportError:
+            raise ImportError(
+                "You must run `pip install --upgrade "
+                "google-cloud-aiplatform google-cloud-storage`"
+                "to use the MatchingEngine Vectorstore."
+            )
+
+    def _validate_gcs_bucket(self, gcs_bucket_uri: str) -> str:
+        """Validates the gcs_bucket_uri as a bucket name.
+
+        Args:
+              gcs_bucket_uri: The received bucket uri.
+
+        Returns:
+              A valid gcs_bucket_uri or throws ValueError if full path is
+              provided.
+        """
+        gcs_bucket_uri = gcs_bucket_uri.replace("gs://", "")
+        if "/" in gcs_bucket_uri:
+            raise ValueError(f"The argument gcs_bucket_uri should only be "
+                             f"the bucket name. Received {gcs_bucket_uri}")
+        return gcs_bucket_uri
+
+    @classmethod
+    def _create_credentials_from_file(
+        cls,
+        json_credentials_path: Optional[str]
```

Comment:
Looks like this is only called from the other classmethod in which the creds path is required. We may as well make it required here then

Reply:
Actually, I ended up adding `Optional[str]` to `from_components` because the idea is that if the user doesn't pass any credentials, then the default will be used.